File: ContiguousArr.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();

        int longest = 0;
        int rsum = 0;

        //initialize the value for 0th element to factor in the first element in array
        map.put(0, -1);

        for(int i = 0; i < nums.length; i++){
            if(nums[i] == 0)
                rsum = rsum - 1;
            else
                rsum = rsum + 1;

            //if map contains the running sum, check if its the longest
            //if doesnt exist, add to it with the index value
            if(map.containsKey(rsum))
                longest = Math.max(longest, i - map.get(rsum));
            else
                map.put(rsum, i);
        }

        return longest;
    }
}

File: ContiguousArr.cpp
// Time Complexity : O(n)
// Space Complexity : O(n) as we maintain a hashmap
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No
#include<unordered_map>
#include<vector>
using namespace std;
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        //key will be  unique running sum and value will be earliest index 
        unordered_map<int,int> uniqueSum;
        int maxLen =  0;
        int rSum=0;
        int i =-1;
        uniqueSum[rSum] = i;
        i++;
        for(i;i<nums.size();i++)
        {
            nums[i]?rSum++:rSum--;
            if(uniqueSum.find(rSum)!=uniqueSum.end())
            {
                maxLen = max(maxLen,(i-uniqueSum[rSum]));
            }
            else
            {
                uniqueSum[rSum]=i;
            }
        }
        return maxLen;
    }
};